### PR TITLE
Add 'types' script module for checking Defold userdata types.

### LINKED ISSUE #8071: Functions to identify hash, vector3 and other `userdata` types
**Is your feature request related to a problem? Please describe (REQUIRED):**
Data type checking is often needed for convenient code organisation. Unfortunately, `type(variable)` always returns `userdata` for hash, vector3 and other types built into Defold.

```lua
    assert(type(self.position) == "userdata")
    assert(type(self.body_id) == "userdata")
    assert(type(self.collision_url) == "userdata")
    assert(type(self.rotation) == "userdata")
```

You have to write it that way and guess what's inside the variables.

**Describe the solution you'd like (REQUIRED):**
It would be much more understandable if you could write it like this:

```lua
    assert(types.is_vector3(self.position))
    assert(types.is_hash(self.body_id))
    assert(types.is_url(self.collision_url))
    assert(types.is_quat(self.rotation))
```

**Describe alternatives you've considered (REQUIRED):**
I have added similar features for Scene3D and it has proven to be very handy in game development.
https://github.com/indiesoftby/defold-scene3d/blob/main/scene3d/src/extension.cpp#L12

### PR DESCRIPTION
**Pull request by ekharkunov**
Changes added new Lua module '**types**' which contains utility function to check Defold userdata type like vector, vector3, matrix, quaternion, etc. Module can be excluded via AppManifest.

Usage example:
```lua
    local url = msg.url("main_collection", "/hero", "right_hand")
    print(types.is_url(url)) ----> true

    local long_vector = vmath.vector({ 10, 14, 19, -20, -11, -2, 132 })
    print(types.is_vector(long_vector)) --------> true

    local non_empty_vector3 = vmath.vector3(13, 14.5, -7)
    print(types.is_vector3(non_empty_vector3)) ------> true

    local non_empty_matrix = vmath.matrix4_translation(non_empty_vector3)
    print(types.is_matrix4(non_empty_matrix)) -----> true

    local non_empty_quat = vmath.quat(1, 2, 3, 4)
    print(types.is_quat(non_empty_quat)) ------> true

    local test_hash = hash("some_foo_text")
    print(types.is_hash(test_hash))  --- ---> true

    print(not types.is_vector(non_empty_vector3))   ------> false
    print(not types.is_vector(nil))   ------> false
    print(not types.is_vector3(non_empty_matrix))   ------> false
    print(not types.is_quat(non_empty_vector3))   ------> false
    print(not types.is_url(long_vector))   ------> false
    print(not types.is_matrix4(test_hash))   ------> false
```

Fixes #8071 

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [x] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [x] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [x] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

**Comment by aglitchman** [comment]
Fantastic!!! 🤩

File: engine/script/src/script_types.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+namespace dmScript
+{
+    /*# Types API documentation
+     *
+     * Functions for checking Defold userdata types.
+     *
+     * @document
+     * @name Types
+     * @namespace types
```
Comment (AGulev) [inline engine/script/src/script_types.cpp:34]: pay attention to lua.clj file 
https://github.com/defold/defold/blob/de8e81e893ced421e2a820eeeec70d78675f6212/editor/src/clj/editor/lua.clj#L32-L38

and

https://github.com/defold/defold/blob/de8e81e893ced421e2a820eeeec70d78675f6212/engine/docs/src/script_doc.py#L497-L518

**Comment by Jhonnyg** [comment]
Quick question:
Could the types be part of the metatable of the type userdatum themselves? And do we really need to exclude the type library? 🤔

**Comment by ekharkunov** [comment]
> Quick question: Could the types be part of the metatable of the type userdatum themselves?And do we really need to exclude the type library? 🤔
> 
> edit: type isn’t part of the meta table it seems

If I understand everything correctly custom type is already part of userdata
https://github.com/defold/defold/blob/dev/engine/script/src/script.cpp#L737-L766

Regarding excluding module: I make in such way because it was concern about engine from Bjorn's side: https://github.com/defold/defold/issues/8071#issuecomment-1946351209

**Review by JCash — APPROVED** [review]

**Pull request merged**